### PR TITLE
Preserve non-reserved metadata

### LINKED ISSUE #438: Preserve non-reserved metadata entries

### PR DESCRIPTION
**Pull request by robfletcher**
Fixes #438

### CODE REVIEW

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ResourceController.kt [hunk 1]
```diff
@@ -94,9 +94,9 @@ class ResourceController(
     log.error(e.message)
   }
 
-  @ExceptionHandler(HttpMessageNotReadableException::class)
+  @ExceptionHandler(HttpMessageConversionException::class)
```
Comment (robfletcher) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ResourceController.kt]: This is just a broader exception. If you have validation rules in the `init` block of the data class it throws the more general exception.

File: keel-api/src/test/kotlin/com/netflix/spinnaker/keel/rest/ResourceControllerTests.kt [hunk 2]
```diff
@@ -211,8 +211,7 @@ internal class ResourceControllerTests {
           |apiVersion: test.spinnaker.netflix.com/v1
           |kind: whatever
           |metadata:
-          |  serviceAccount: keel@spinnaker
-          |  name: i-should-not-be-naming-my-resources-that-is-keels-job
+          |  name: i-forgot-my-service-account
```
Comment (robfletcher) [inline keel-api/src/test/kotlin/com/netflix/spinnaker/keel/rest/ResourceControllerTests.kt]: Changed the intent of this test

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Resource.kt [hunk 3]
```diff
@@ -73,18 +73,15 @@ data class SubmittedResource<T : ResourceSpec>(
     property = "kind"
   )
   val spec: T
-)
+) {
+  init {
+    require(metadata["serviceAccount"].isValidServiceAccount()) { "serviceAccount must be a valid service account" }
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Resource.kt]: require a service account on submission

File: keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResolvableResourceHandler.kt [hunk 4]
```diff
@@ -45,10 +45,9 @@ interface ResolvableResourceHandler<S : ResourceSpec, R : Any> : KeelPlugin {
    * applied, etc.
    */
   fun normalize(resource: SubmittedResource<S>): Resource<S> {
-    val metadata = mapOf(
+    val metadata = resource.metadata + mapOf(
```
Comment (robfletcher) [inline keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResolvableResourceHandler.kt]: make sure we keep any user-supplied metadata

**Comment by emjburns** [comment]
Does this serialize correctly in/out of the sql database?

**Comment by robfletcher** [comment]
@emjburns we never put `SubmittedResource` into the DB. This change doesn't affect `Resource`.

**Comment by emjburns** [comment]
LOL obviously I knew that GREAT. LGTM

**Pull request merged**